### PR TITLE
[stable/redis-ha] Support topologySpreadConstraints

### LINKED ISSUE #128: [chart/redis-ha][REQUEST] support topologySpreadConstraints
**Is your feature request related to a problem? Please describe.**
I want to modify topologySpreadConstraints in values.yaml. (Pod Topology Spread Constraints feature - k8s 1.19 GA / previous version is Alpha or Beta)
https://kubernetes.io/docs/concepts/workloads/pods/pod-topology-spread-constraints/

I am using two nodes and always have to distribute pods to two nodes as "1 / 2".
In order to do that, I have used the requiredDuringSchedulingIgnoredDuringExecution and preferredDuringSchedulingIgnoredDuringExecution options in podAntiAffinity, but both of them have not met the condition and have to use topologySpreadConstraints.

**Describe the solution you'd like**
I thought about two ways.
Label is determined in helm chart, i think first thing seems better

1) just enabled topologySpreadConstraints and set values
```
      {{- if .Values.topologySpreadConstraints.enabled }}
      topologySpreadConstraints:
      - maxSkew: {{ .Values.topologySpreadConstraints.maxSkew }}
        topologyKey: {{ .Values.topologySpreadConstraints.topologyKey }}
        whenUnsatisfiable: {{ .Values.topologySpreadConstraints.whenUnsatisfiable }}
        labelSelector:
          matchLabels:
            release: {{ .Release.Name }}
            app: {{ template "redis-ha.name" . }}
      {{- end }}
```

2) get values from values.yaml file
```
      {{- if .Values.topologySpreadConstraints }}
      topologySpreadConstraints:
{{ toYaml .Values.topologySpreadConstraints | indent 8 }}
      {{- end }}
```

Thank you!

### PR DESCRIPTION
**Pull request by TheRealDwright**
Closes #128

#### Which issue this PR fixes
  - fixes #128

#### Special notes for your reviewer:

Topology spread will be soft enforcement by default so should not have any surprising consequences
to those who choose to enable it. This can be a hard enforcement by setting whenUnsatisfiable to
`DoNotSchedule`

#### Checklist
[Place an '[x]' (no spaces) in all applicable fields. Please remove unrelated fields.]
- [x] [DCO](https://github.com/helm/charts/blob/master/CONTRIBUTING.md#sign-your-work) signed
- [x] Chart Version bumped
- [x] Variables are documented in the README.md
- [x] Title of the PR starts with chart name (e.g. `[stable/mychartname]`)

### CODE REVIEW

**Comment by DandyDeveloper** [comment]
@TheRealDwright Sorry, I've neglected this PR, the autoassign isn't working. I'll take a look and test these constraints when I have a free moment.

**Comment by TheRealDwright** [comment]
> @TheRealDwright Sorry, I've neglected this PR, the autoassign isn't working. I'll take a look and test these constraints when I have a free moment.

No problem, we've already tested this rigorously in our production environment, but feel free to do your due diligence and let me know if you want me to update it with any changes.

**Comment by TheRealDwright** [comment]
Hey @DandyDeveloper - is there anything you would like me to do/change on this?

File: charts/redis-ha/values.yaml [hunk 1]
```diff
@@ -430,6 +430,10 @@ additionalAffinities: {}
 ##
 affinity: |
 
+## https://kubernetes.io/docs/concepts/workloads/pods/pod-topology-spread-constraints/
+topologySpreadConstraints:
+  enabled: false
+
```
Comment (DandyDeveloper) [inline charts/redis-ha/values.yaml:436]: Can we have all the options visible in the values? 

```suggestion
topologySpreadConstraints:
  enabled: false
  maxSkew: ""
  topologyKey: ""
  whenUnsatisfiable: "" 
```

With this, think we're good to merge in.

**Review by DandyDeveloper — CHANGES_REQUESTED** [review]

**Review by DandyDeveloper — APPROVED** [review]

**Comment by mkilchhofer** [comment]
Hmm maybe we should combine this with `.Values.hardAntiAffinity`?

I think it would make sense to disable the affinity completely when using topologySpreadConstraints. WDYT @TheRealDwright and @DandyDeveloper ?

**Comment by DandyDeveloper** [comment]
@mkilchhofer Technically, both can be used in conjunction to still define hard scheduling requirements for pods. Best "overview" of this I can find would be this: https://kubernetes.io/blog/2020/05/introducing-podtopologyspread/#usage-along-with-nodeselector-nodeaffinity

I don't think the two should be connected.

**Comment by TheRealDwright** [comment]
Agree with @DandyDeveloper - all this PR is intended to facilitate is the even distribution of pods between cluster zones (e.g. AWS Availability Zones) for high availability purposes. It's only loosely related to anti-affinity.

**Pull request merged**